odoo/addons/base/models/ir_qweb.py

```diff
@@ -2389,43 +2533,31 @@ def _compile_directive_cache(self, el, compile_context, level):
-            cache_key = self._get_cache_key(template_cache_key) if template_cache_key else None
-            uniq_cache_key = cache_key and ({str(self.env.context['__qweb_base_key_cache'])!r}, '{def_name}_cache', cache_key)
-            loaded_values = values['__qweb_loaded_values']
-            def {def_name}_cache():
-                content = []
-                text = []
-                for item in {def_name}(self, {{**values, '__qweb_in_cache': True}}):
-                    if isinstance(item, str):
-                        text.append(item)
-                    else:
-                        content.append(''.join(text))
-                        content.append(item)
-                        text = []
-                if text:
-                    content.append(''.join(text))
-                return content
-            cache_content = self._load_values(uniq_cache_key, {def_name}_cache, loaded_values)
-            if values.get('__qweb_in_cache'):
-                yield from cache_content
-            else:
-                for item in cache_content:
-                    if isinstance(item, str):
-                        yield item
-                    else:
-                        ref, function_name, cached_values = item
-                        t_nocache_function = loaded_values.get(function_name)
-                        if not t_nocache_function:
-                            t_call_template_functions, def_name = self._compile(ref)
-                            t_nocache_function = t_call_template_functions[function_name]
-
-                        nocache_values = values['__qweb_root_values'].copy()
-                        nocache_values.update(cached_values)
-                        yield ''.join(t_nocache_function(self, nocache_values))
-            """, level))
-
+        def_wrap = [indent_code(f"""
+            def {def_name}_wrap(self, values):
+                # element: {path!r} , {xml!r}'
+                template_cache_key = {self._compile_expr(expr)} if not self.env.context.get('is_t_cache_disabled') else None
+                cache_key = self._get_cache_key(template_cache_key) if template_cache_key else None
+                uniq_cache_key = cache_key and ({str(self.env.context['__qweb_base_key_cache'])!r}, '{def_name}_cache', cache_key)
+                def {def_name}_cache():
+                    text = []
+                    result = []
+                    for item in self.with_context(__qweb_dont_fetch_directive='t-nocache')._render_iterall({compile_context['template']!r}, {def_name!r}, values):
+                        if isinstance(item, str):
+                            text.append(item)
+                        else:
+                            if text:
+                                result.append(''.join(text))
+                                text = []
+                            result.append(item)
+                    if text: result.append(''.join(text))
+                    return result
```

Comment:
Nitpick. Is this equivalent?
```suggestion
                    return list(
                      item
                      for item_type, items in itertools.groupby(self.with_context(__qweb_dont_fetch_directive='t-nocache')._render_iterall({compile_context['template']!r}, {def_name!r}, values), type)
                      for item in ((''.join(items),) if issubtype(item_type, str) else items)
                    )
```

Actually, you could just return the `list(self.with_context..._render_iterall(...))` and let the parent do the join. Maybe you already tested the performance and that would be slower.

Reply:
It will add 'itertools' in the qweb globals. We should have the minimal set of globals.